x-pack/plugins/licensing/public/plugin.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { BehaviorSubject, Subject, merge, of, timer } from 'rxjs';
+import { filter, map, pairwise, switchMap, takeUntil, tap } from 'rxjs/operators';
+import {
+  CoreSetup,
+  CoreStart,
+  Plugin as CorePlugin,
+  PluginInitializerContext,
+} from 'src/core/public';
+import { LicensingPluginSetup, ILicensingPlugin, ObjectifiedLicense } from '../common/types';
+import {
+  API_ROUTE,
+  LICENSING_SESSION,
+  LICENSING_SESSION_SIGNATURE,
+  SIGNATURE_HEADER,
+  DEFAULT_POLLING_FREQUENCY,
+} from '../common/constants';
+import { License } from '../common/license';
+import { hasLicenseInfoChanged } from '../common/has_license_info_changed';
+
+export class Plugin implements CorePlugin<LicensingPluginSetup>, ILicensingPlugin {
+  private stop$ = new Subject();
+  private removeInterceptor!: () => void;
+  private signature = '';
+  public pollingFrequency = DEFAULT_POLLING_FREQUENCY;
+  private refresher$ = new BehaviorSubject(true);
+
+  constructor(context: PluginInitializerContext) {}
+
+  private getSession(): ObjectifiedLicense | null {
+    const raw = sessionStorage.getItem(LICENSING_SESSION);
+    const signature = sessionStorage.getItem(LICENSING_SESSION_SIGNATURE);
+    const json = raw && JSON.parse(raw);
+
+    if (signature) {
+      this.signature = signature;
+    }
+
+    return json;
+  }
+
+  private setSession = (license: License) => {
+    sessionStorage.setItem(LICENSING_SESSION, JSON.stringify(license.toObject()));
+
+    if (license.signature) {
+      sessionStorage.setItem(LICENSING_SESSION_SIGNATURE, license.signature);
+    }
+  };
+
+  private clearSession() {
+    sessionStorage.removeItem(LICENSING_SESSION_SIGNATURE);
+    sessionStorage.removeItem(LICENSING_SESSION);
+  }
+
+  public sign() {
+    return this.signature;
+  }
+
+  public setup(core: CoreSetup) {
+    const session = this.getSession();
+    const initial$ = of(
+      session
+        ? License.fromObjectified(session, { plugin: this })
+        : new License({ plugin: this, features: {} })
+    );
+    const setup = {
+      refresh: () => this.refresher$.next(true),
+      license$: initial$,
+    };
+
+    this.removeInterceptor = core.http.intercept({
+      response: httpResponse => {
+        const signature =
+          (httpResponse.response && httpResponse.response.headers.get(SIGNATURE_HEADER)) || '';
+
+        if (signature !== this.signature) {
+          this.signature = signature;
+
+          if (httpResponse.request && !httpResponse.request.url.includes(API_ROUTE)) {
```

Comment:
License plugin in the LP intercepts all the request and check the `signature` header.
In the NP we mix this logic + polling via intervals. Why? That complicates things without much benefit.

Reply:
> In the NP we mix this logic + polling via intervals.

This statement is inaccurate. The interceptor lives outside of both polling and intervals, so the logic has not really changed, only using the NP interceptor implementation.

Reply:
it's still doesn't answer why we need polling mechanism

Reply:
We use polling to check for changes to the license.

Reply:
Sure. The question: why? We already have the request interceptor for this. Why was it sufficient in the LP, but not enough in the NP?

Reply:
It wasn't sufficient in the legacy platform either, we polled for it on the server, but not the client, and for consistency we want both client and server to get an update to the license as soon as we can detect it.

Reply:
I think there's a few things here we have to consider:
- Each open Kibana tab would now be polling when it wasn't before.
- We've had issues in the past on Cloud with Kibana's different pollers being too frequent and causing strain on load balancers.
- Cloud customers pay for data transfers and this would increase how much data in transferred (though probably (?) a negligible mount)

Are there known issues around this in the legacy implementation? I wasn't aware that it was different in the legacy plugin, but it makes sense to piggyback on HTTP interceptors, if we can, to reduce request load. 

In the future, we can save a lot of time if this difference was highlighted in the PR description so we discuss high-level details first, before spending time reviewing details that may be eliminated by a design change.

This issue was meant to be a pretty simple migration from Legacy to New Platform. If we are making design changes, those need to be documented and discussed to ensure that these decisions are well informed by the requirements of the product. At Elastic, we put the onus of that process on the developer. Catching this change this late in the process is putting our timeline in jeopardy.